### PR TITLE
sphinx gallery

### PR DESCRIPTION
**Pull request by clewis7**
start #486 for hopefully the last time :)

According to @almarklein, should just be able to directly use `pygfx image scraper`

Initially, I am just trying to get it to work with one set of the desktop examples.

Current issues: 
1. Offscreen rendering is being ignored even if it is in `os.environ` keys, should have maybe been fixed by #431
2. When using `gflw WgpuCanvas`, the `pygfx` scraper is looking for the `canvas` or `renderer` and then attempting to call `canvas.draw()` but the desired method is `canvas.draw_frame()`
3. Beyond this, if I in-place change the source code of the `pygfx` scraper to use the correct draw method, an issue arises with trying to write a frame:

```
    Traceback (most recent call last):
      File "/home/caitlin/venvs/fpl/lib/python3.12/site-packages/sphinx_gallery/scrapers.py", line 341, in save_figures
        rst = scraper(block, block_vars, gallery_conf)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/caitlin/venvs/fpl/lib/python3.12/site-packages/pygfx/utils/gallery_scraper.py", line 136, in pygfx_scraper
        render_screenshot(canvas, namespace, img_filename, **config)
      File "/home/caitlin/venvs/fpl/lib/python3.12/site-packages/pygfx/utils/gallery_scraper.py", line 209, in render_screenshot
        iio.imwrite(
      File "/home/caitlin/venvs/fpl/lib/python3.12/site-packages/imageio/v3.py", line 147, in imwrite
        encoded = img_file.write(image, **kwargs)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/caitlin/venvs/fpl/lib/python3.12/site-packages/imageio/plugins/pillow.py", line 443, in write
        for frame in ndimage:
    TypeError: iteration over a 0-d array
```

@kushalkolar @almarklein thoughts?

### CODE REVIEW

File: examples/desktop/gridplot/gridplot.py
Comment (clewis7) [inline examples/desktop/gridplot/gridplot.py:14]: As noted in `pygfx` docs, scraper looks for canvas or renderer...I think examples will need to all have `canvas` variable defined

Comment (kushalkolar) [reply examples/desktop/gridplot/gridplot.py:14]: canvas = fig.canvas ?

Comment (clewis7) [reply examples/desktop/gridplot/gridplot.py:14]: Currently, I feel like the biggest issue is that no matter how I pass the canvas, when the docs build fpl is always forcing use of a `GlfwWgpuCanvas` instead of `WgpuCanvas`

Comment (clewis7) [reply examples/desktop/gridplot/gridplot.py:14]: working solution, force the canvas like so:

```
from wgpu.gui.offscreen import WgpuCanvas

canvas = WgpuCanvas(size=(800, 800))
```

Comment (kushalkolar) [reply examples/desktop/gridplot/gridplot.py:14]: This won't be a useable example then. Check the selection logic in the gui module

Comment (clewis7) [reply examples/desktop/gridplot/gridplot.py:14]: I am so silly, just need to set the `os.environ` variable before `fastplotlib` gets imported so the gui autodetects the correct thing!!

Comment (kushalkolar) [reply examples/desktop/gridplot/gridplot.py:14]: so you can do that in the sphinx `conf.py` ?

Comment (clewis7) [reply examples/desktop/gridplot/gridplot.py:14]: yes!!

**Comment by clewis7** [comment]
Can now force offscreen rendering, so the gallery is being generated.

Current issue is figuring out how to format the `index.rst` and `conf.py` in order to properly display the gallery.

1) For some reason the dropdown for all the examples only gets placed on the last one
image
2) The thumbnails are being displayed weird
image

File: docs/source/quickstart.ipynb
Comment (kushalkolar) [inline docs/source/quickstart.ipynb:1]: I think that instead you can add to our `gui` module that if env var `NB_SNAPSHOT=1`, force jupyter canvas?

Comment (clewis7) [reply docs/source/quickstart.ipynb:1]: I think the only issue with this, is that we would still need a cell in the quickstart to set that environment variable because if we set it before in `conf.py` for the doc build it will attempt to use jupyter canvas for the gallery

Comment (clewis7) [reply docs/source/quickstart.ipynb:1]: I tried this, and gallery gets run first before the nbsphinx plugin...so we would either need to keep the cell where we turn off the offscreen rendering in quickstart.ipynb or add the logic you talked about above

**Comment by clewis7** [comment]
Fixed dropdown issue from above, still working on issues with thumbnails appearing weird and thumbnail links being broken

I think because our examples structure has the extra layer of `desktop` folder it is making things generate wonky

**Comment by BalzaniEdoardo** [comment]
@clewis7 remove the "nbsphinx", it seems to create some conflicts

https://github.com/sphinx-gallery/sphinx-gallery/issues/1195

File: docs/source/conf.py
Comment (BalzaniEdoardo) [inline docs/source/conf.py:41]:
```suggestion
```

**Comment by clewis7** [comment]
@kushalkolar Like Edo said, in order to use `sphinx-gallery` we would need to remove `nbsphinx`. Pitfall of this is that then the `quickstart.ipynb` will not render...

**Comment by kushalkolar** [comment]
This is great :smile: !

Some things:

Can you pre-download the images in `conf.py` so we dont get this:

image

Also need to figure out how to properly set the canvas size so that when the screenshots are taken we don't have that huge blank space. 

Can you check how we can have animated examples, like pygfx: https://pygfx.com/stable/_gallery/introductory/cube.html#sphx-glr-gallery-introductory-cube-py 

We can show some of the animations from quickstart, and I guess if you tag it with "test_examples = false" it will generate sphinx gallery docs with that example but not use it for the test suite? Since we don't want animation examples to run with screenshot tests (at least not yet, maybe later :smile: )

File: docs/source/conf.py
Comment (kushalkolar) [inline docs/source/conf.py:9]: comment

File: docs/source/conf.py
Comment (kushalkolar) [inline docs/source/conf.py:18]: use pathlib.joinpath

File: examples/desktop/gridplot/gridplot.py
Comment (kushalkolar) [inline examples/desktop/gridplot/gridplot.py:26]: specifiy that this line isn't required for users, or see if there's a way for the `pygfx` scraper to look for `fig.canvas` instead of `canvas`? 

in all examples

File: examples/desktop/heatmap/heatmap.py
Comment (kushalkolar) [inline examples/desktop/heatmap/heatmap.py:16]: remove

Comment (kushalkolar) [reply examples/desktop/heatmap/heatmap.py:16]: this example has changed anyways in #511 so let's wait until that is merged, and then rebase this on that.

File: examples/desktop/heatmap/heatmap_cmap.py
Comment (kushalkolar) [inline examples/desktop/heatmap/heatmap_cmap.py:16]: see previous

File: fastplotlib/layouts/_figure.py
Comment (kushalkolar) [inline fastplotlib/layouts/_figure.py:483]: since we're not using `nbsphinx` we can get rid of all this

File: setup.py
Comment (kushalkolar) [inline setup.py:26]: ??????????

**Review by kushalkolar — CHANGES_REQUESTED** [review]
More:

For each example, specify that the `if __name__ == "__main__"` is NOT how it's supposed to be done for interactive use. We should make a doc page on interactive use in `ipython` and `jupyter` using this: https://wgpu-py.readthedocs.io/en/stable/gui.html#using-wgpu-interactively 

Users should never call `fpl.run()` for interactive use.

File: examples/desktop/gridplot/gridplot.py
Comment (clewis7) [reply examples/desktop/gridplot/gridplot.py:26]: I think perhaps we can use `figure` instead...

`pygfx` scraper looks for one of `["renderer", "canvas"]` to be defined, but will also look through proxy objects such as `"figure", "display", "disp", "plot"` 

So I will try that

File: examples/desktop/heatmap/heatmap.py
Comment (clewis7) [reply examples/desktop/heatmap/heatmap.py:16]: I'll just remove the heatmap examples from being generated for now

File: setup.py
Comment (clewis7) [reply setup.py:26]: sphinx-gallery requires matplotlib, and need imageio for the examples

File: examples/desktop/gridplot/gridplot.py
Comment (clewis7) [reply examples/desktop/gridplot/gridplot.py:26]: I think for now, we should just to `canvas = fig.canvas` and have a note saying that it is not required for users and after FENS I can look into fixing them

**Comment by clewis7** [comment]
**NOTE:**
sphinx-gallery doesn't have the namespace `__file__`, so in the `scatter/` examples will need to change how the path to the `iris.npy` data is loaded so that gallery can be generated properly and tests will still run without using `__file__` or `os.getcwd()`

**Comment by kushalkolar** [comment]
After this I propose:

1. get rid of quickstart.ipynb, replace with a guide in the docs
2. get rid of any example nbs that are in the examples gallery, except for the selector tools? 
3. what about moving all example nbs to the examples gallery and having just a few example nbs? Like ImageWidget, selector tools, and maybe one simple one that shows layouting with VBox/HBox and a few graphics?

**Comment by clewis7** [comment]
> After this I propose:
> 
> 1. get rid of quickstart.ipynb, replace with a guide in the docs
> 2. get rid of any example nbs that are in the examples gallery, except for the selector tools?
> 3. what about moving all example nbs to the examples gallery and having just a few example nbs? Like ImageWidget, selector tools, and maybe one simple one that shows layouting with VBox/HBox and a few graphics?

This sounds good to me, and doable before FENS

**Comment by kushalkolar** [comment]
A few event examples too, click event that changes colors of closest circle, and another example where changing a feature prints the event attributes and event.info.

File: examples/desktop/gridplot/multigraphic_gridplot.py
Comment (kushalkolar) [inline examples/desktop/gridplot/multigraphic_gridplot.py:1]: put an image under the circles, and use an alpha for the circles

**Comment by clewis7** [comment]
I added a simple event example but I am going to hold off adding any more examples until we merge this with graphic features PR

**Comment by clewis7** [comment]
@kushalkolar merged main and updated...should be ready for review :D

File: examples/desktop/screenshots/heatmap.png
Comment (kushalkolar) [inline examples/desktop/screenshots/heatmap.png:1]: you'll need to remove the old screenshots and add the new ones, make sure they look like the ones from gfeatures refactor but just cropped

File: fastplotlib/layouts/_figure.py
Comment (kushalkolar) [reply fastplotlib/layouts/_figure.py:482]: seems like this came back??

File: examples/desktop/scatter/scatter_iris.py
Comment (kushalkolar) [inline examples/desktop/scatter/scatter_iris.py:8]: don't we want this in the tests but not the docs?

File: examples/desktop/scatter/scatter_dataslice.py
Comment (kushalkolar) [inline examples/desktop/scatter/scatter_dataslice.py:8]: I thought we want this in docs only but not tests since it creates random points

File: examples/desktop/misc/simple_event.py
Comment (kushalkolar) [inline examples/desktop/misc/simple_event.py:2]: This should be much simpler, only one graphic, change the colors, two callbacks one decorated and one not

Comment (kushalkolar) [reply examples/desktop/misc/simple_event.py:2]: Also add an example of click events, maybe with an image

**Review by kushalkolar — CHANGES_REQUESTED** [review]

**Comment by kushalkolar** [comment]
This example seems like it's missing from the gallery :thinking: https://github.com/fastplotlib/fastplotlib/blob/main/examples/desktop/line_collection/line_stack_3d.py

File: examples/desktop/gridplot/multigraphic_gridplot.py
Comment (clewis7) [reply examples/desktop/gridplot/multigraphic_gridplot.py:112]: #510

**Comment by clewis7** [comment]
> This example seems like it's missing from the gallery 🤔 https://github.com/fastplotlib/fastplotlib/blob/main/examples/desktop/line_collection/line_stack_3d.py

ahhh, wasn't rendering properly because of the canvas size

I found that the canvas size needs to be `(700, 560)`...`sphinx-gallery` likes that best

**Comment by clewis7** [comment]
yayyyy! @kushalkolar

File: examples/desktop/misc/simple_event.py
Comment (kushalkolar) [inline examples/desktop/misc/simple_event.py:37]:
```suggestion
# adding a click event, we can also use decorators to add event handlers
```

File: examples/desktop/scatter/scatter_cmap_iris.py
Comment (kushalkolar) [inline examples/desktop/scatter/scatter_cmap_iris.py:20]:
```suggestion
data = datasets.load_iris()["data"]
```

File: examples/desktop/scatter/scatter_cmap_iris.py
Comment (kushalkolar) [inline examples/desktop/scatter/scatter_cmap_iris.py:15]:
```suggestion
from sklearn import datasets
```

File: examples/desktop/scatter/scatter_cmap_iris.py
Comment (kushalkolar) [inline examples/desktop/scatter/scatter_cmap_iris.py:9]:
```
# sphinx_gallery_pygfx_docs = 'screenshot'
```

File: examples/desktop/scatter/scatter_cmap_iris.py
Comment (kushalkolar) [inline examples/desktop/scatter/scatter_cmap_iris.py:1]: by adding sklearn to docs deps we can just load the iris dataset, it's a nice and simple usecase example!

**Review by kushalkolar — APPROVED** [review]

**Pull request merged**
